kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ProxyReconciler.java

```diff
@@ -5,28 +5,192 @@
+    public static final String CLUSTERS_DEP = "clusters";
+
+    @Override
     public UpdateControl<KafkaProxy> reconcile(KafkaProxy primary,
                                                Context<KafkaProxy> context) {
         return UpdateControl.patchStatus(
-                new KafkaProxyBuilder(primary).editOrNewStatus()
-                        .withObservedGeneration(primary.getMetadata().getGeneration())
-                        .endStatus()
-                        .build());
+                buildStatus(primary, null));
+    }
+
+    @Override
+    public ErrorStatusUpdateControl<KafkaProxy> updateErrorStatus(KafkaProxy primary,
+                                                                  Context<KafkaProxy> context,
+                                                                  Exception exception) {
+        // Post-condition: status.conditions should be in a canonical order (to avoid non-terminating reconciliations)
+        // Post-condition: There is only one Ready condition
+        var control = ErrorStatusUpdateControl.patchStatus(buildStatus(primary, exception));
+        if (exception instanceof InvalidResourceException) {
+            control.withNoRetry();
+        }
+        else {
+            control.rescheduleAfter(Duration.ofSeconds(10));
+        }
+        return control;
+
+    }
+
+    private static KafkaProxy buildStatus(KafkaProxy primary,
+                                          @Nullable Exception exception) {
+        // @formatter:off
+        return new KafkaProxyBuilder(primary)
+                .editOrNewStatus()
+                    .withObservedGeneration(primary.getMetadata().getGeneration())
+                    .withConditions(effectiveReadyCondition(primary, exception))
+                .endStatus()
+            .build();
+        // @formatter:on
+    }
+
+    /**
+     * Determines whether the {@code Ready} condition has had a state transition,
+     * and returns an appropriate {@code Ready} condition.
+     * @param primary The primary.
+     * @param exception An exception, or null if the reconciliation was successful.
+     * @return The {@code Ready} condition to use in {@code status.conditions}.
+     */
+    private static Conditions effectiveReadyCondition(KafkaProxy primary,
+                                                      @Nullable Exception exception) {
+        final var oldReady = primary.getStatus() == null || primary.getStatus().getConditions() == null
+                ? null
+                : primary.getStatus().getConditions().stream().filter(c -> "Ready".equals(c.getType())).findFirst().orElse(null);
+
+        Conditions newReady = newReadyCondition(primary, exception);
+
+        final boolean useNew;
+
+        if (oldReady == null) {
+            useNew = true;
+        }
+        else if (Conditions.Status.TRUE.equals(oldReady.getStatus()) && exception == null) {
+            // Both Ready=True
+            useNew = false;
```

Comment:
Something feels off to me about having both `=true` and `=false` here. If both old and new are ready why do we prefer old?

Reply:
I wanted to guarantee that the `lastTransitionTime` only changed when there _really was_ a transition (i.e. I wanted to avoid it getting updated on every reconciliation if the rest of it was the same). I've rewritten this to make that more obvious (adding an `isTransition()` method, a call to which guards the creation of a `newReady`).